graph/src/blockchain/firehose_block_ingestor.rs

```diff
@@ -192,3 +157,69 @@ where
         Ok(())
     }
 }
+
+#[async_trait]
+impl<M, C: Blockchain> BlockIngestor for FirehoseBlockIngestor<M, C>
+where
+    M: prost::Message + BlockchainBlock + Default + 'static,
+{
+    async fn run(self: Box<Self>) {
+        let mut latest_cursor = self.fetch_head_cursor().await;
+        let mut backoff =
+            ExponentialBackoff::new(Duration::from_millis(250), Duration::from_secs(30));
+
+        loop {
+            let endpoint = match self.client.firehose_endpoint() {
+                Ok(endpoint) => endpoint,
+                Err(err) => {
+                    error!(
+                        self.logger,
+                        "Unable to get a connection for block ingestor, err: {}", err
+                    );
+                    backoff.sleep_async().await;
+                    continue;
+                }
+            };
+
+            let logger = self.logger.new(
+                o!("provider" => endpoint.provider.clone(), "network_name"=> self.network_name()),
+            );
+
+            info!(
+                logger,
+                "Blockstream disconnected, connecting"; "endpoint uri" => format_args!("{}", endpoint), "cursor" => format_args!("{}", latest_cursor),
+            );
+
+            let result = endpoint
+                .clone()
+                .stream_blocks(firehose::Request {
+                    // Starts at current HEAD block of the chain (viewed from Firehose side)
+                    start_block_num: -1,
+                    cursor: latest_cursor.clone(),
+                    final_blocks_only: false,
+                    transforms: self.default_transforms.iter().map(|t| t.into()).collect(),
+                    ..Default::default()
+                })
+                .await;
+
+            match result {
+                Ok(stream) => {
+                    info!(logger, "Blockstream connected, consuming blocks");
+
+                    // Consume the stream of blocks until an error is hit
+                    latest_cursor = self.process_blocks(latest_cursor, stream).await
```

Comment:
I know it's not from this PR, but we should have a new type `struct FirehoseCursor(Box<str>)` for passing around firehose cursors instead of just using `String` for clarity. Something for the todo list

Reply:
https://github.com/graphprotocol/graph-node/issues/4419